sql: panic in lex.TestEncodeSQLBytes when given invalid inputs

Found via

```
 make testlogic TESTFLAGS='-verbosity=10 -show-sql -test.v' FILES=cluster_version
...
SET CLUSTER SETTING version = '1.0-500';
panic: runtime error: slice bounds out of range

goroutine 418 [running]:
github.com/cockroachdb/cockroach/pkg/sql/lex.EncodeSQLStringWithFlags(0xc420f59588, 0xc421498520, 0x16, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/lex/encode.go:112 +0x3a3
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DString).Format(0xc42162f6a0, 0xc420f59560)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1120 +0x8b
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0xc420f59560, 0x2787980, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:47 +0xa4
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0xc420f59560, 0x2787980, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:256 +0x383
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).expr(0xc4212c6f80, 0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:364 +0x129
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).(github.com/cockroachdb/cockroach/pkg/sql.expr)-fm(0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:344 +0x73
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).expr(0xc4211da460, 0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:560 +0x8a
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a3080, 0xc4211da050, 0x22c660f, 0x6)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:161 +0x24e
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc4211da460, 0x27a3080, 0xc4211da050, 0x22c77f3, 0x7)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:94 +0x8f
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a3040, 0xc421050690, 0x22c65bb, 0x6)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:439 +0x2d42
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitConcrete(0xc4211da460, 0x27a3040, 0xc421050690)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:107 +0x83
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a2a00, 0xc42039cd20, 0x22c3fe6, 0x5)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:477 +0x396d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc4211da460, 0x27a2a00, 0xc42039cd20, 0x10, 0xc42104dd90)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:94 +0x8f
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0xc42104dd60, 0xc42104dd70, 0xc42104dd80, 0xc42104dd90, 0x6, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:58 +0x193
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).populateEntries(0xc4212c6f80, 0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0x0, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:290 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql.planToString(0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0x0, 0x0, 0x0, 0x27a2a00)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:320 +0x150
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4211a8a38, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:350 +0x5b4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:756 +0xdc6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:401 +0xb82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc4211a8600, 0x27acf40, 0xc4214aa900, 0x0, 0x0, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1135 +0x13cd
github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx.func1(0xc4211a8600, 0x27acf40, 0xc4214aa900, 0xc420421a20, 0xc4216a0f00, 0xc42138dfc0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:210 +0x8a
created by github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:199 +0x275
Makefile:864: recipe for target 'testbaselogic' failed
make: *** [testbaselogic] Error 2

```

Upon further inspection, `TestEncodeSQLBytes` is not resilient to malformed string inputs.

--- a/pkg/sql/lex/encode.go
+++ b/pkg/sql/lex/encode.go
@@ -94,6 +94,9 @@ func EncodeSQLStringWithFlags(buf *bytes.Buffer, in string, flags EncodeFlags) {
 	bareStrings := flags.HasFlags(EncBareStrings)
 	// Loop through each unicode code point.
 	for i, r := range in {
+		if i < start {
+			continue
+		}
 		ch := byte(r)
 		if r >= 0x20 && r < 0x7F {
 			if mustQuoteMap[ch] {
@@ -123,7 +126,9 @@ func EncodeSQLStringWithFlags(buf *bytes.Buffer, in string, flags EncodeFlags) {
 	if quote {
 		buf.WriteByte('\'') // begin 'xxx' string if nothing was escaped
 	}
-	buf.WriteString(in[start:])
+	if start < len(in) {
+		buf.WriteString(in[start:])
+	}
 	if escapedString || quote {
 		buf.WriteByte('\'')
 	}

--- a/pkg/sql/parser/parse_test.go
+++ b/pkg/sql/parser/parse_test.go
@@ -2171,6 +2171,13 @@ CREATE TABLE foo(a CHAR(0))
                          ^
 `,
 		},
+		{
+			`e'\xad'::string`,
+			`invalid UTF-8 byte sequence
+e'\xad'::string
+^
+`,
+		},
 	}
 	for _, d := range testData {
 		_, err := parser.Parse(d.sql)

--- a/pkg/util/stringencoding/string_encoding.go
+++ b/pkg/util/stringencoding/string_encoding.go
@@ -119,7 +119,9 @@ func EncodeEscapedChar(
 		// Make sure this is run at least once in case ln == -1.
 		buf.Write(HexMap[entireString[currentIdx]])
 		for ri := 1; ri < ln; ri++ {
-			buf.Write(HexMap[entireString[currentIdx+ri]])
+			if currentIdx+ri < len(entireString) {
+				buf.Write(HexMap[entireString[currentIdx+ri]])
+			}
 		}
 	} else if ln == 1 {
 		// For single-byte runes, do the same as encodeSQLBytes.

